### ISSUE TITLE
is uuid equals to serviceUUID 

### ISSUE THREAD

**Issue by MtyldZ** [barakataboujreich/react-native-ble-advertise#5]
Hi, I wrote a very similar code to your expo tutorial [repo](https://github.com/barakataboujreich/expo-ble-scan-advertise-example/blob/main/App.js). At the end it works, but I am scanning with no serviceUUID filter added. Which makes the app iterate trough all found ble devices (my app and any other ble device). 

I tried to add a serviceUUID filter to BleManager.scan function but that causes to not find any device.
```
// this works but discovers unnecessary devices 
await BleManager.scan([], 2, true)

// tried with a uuid, couldn't find any device
await BleManager.scan([APP_BLE_SERVICE_UUID], 2, true)
```  

Here is my advertiser code 
```
    BleAdvertise.setCompanyId(parseInt(APP_BLE_COMPANY_ID, 16));
    setIsAdvertising(true);
    const minor = parseInt(user.shortUUID.slice(0, 4), 16);
    const major = parseInt(user.shortUUID.slice(4, 8), 16);
    BleAdvertise.broadcast(APP_BLE_SERVICE_UUID, minor, major)
      .catch((error) => {
        console.log(
          '[useBluetoothAdvertise] broadcast failed: ' + error.message,
        );
        setIsAdvertising(false);
      });
```

NOTE: Also I added a log to BleManagerDiscoverPeripheral event and got this periphral (from a android device running my app). One thing that looked interesting to me was `advertising.serviceUUIDs` which is `[]`. I could be totally wrong thinking that the uuid entered to broadcast function should be inside that list and it can be used to filter devices while scanning. 
```
{
  "advertising": {
    "manufacturerData": {
      "ffff": {
        "bytes": [
          2,
          21,
          51,
          153,
          191,
          190,
          48,
          111,
          68,
          120,
          162,
          216,
          93,
          213,
          71,
          239,
          189,
          76,
          140,
          197,
          221,
          181,
          199,
          0
        ],
        "data": "AhUzmb++MG9EeKLYXdVH771MjMXdtccA",
        "CDVType": "ArrayBuffer"
      }
    },
    "txPowerLevel": -2147483648,
    "serviceData": {},
    "isConnectable": true,
    "serviceUUIDs": [],
    "manufacturerRawData": {
      "bytes": [
        0,
        0,
        255,
        255,
        2,
        21,
        51,
        153,
        191,
        190,
        48,
        111,
        68,
        120,
        162,
        216,
        93,
        213,
        71,
        239,
        189,
        76,
        140,
        197,
        221,
        181,
        199,
        0
      ],
      "data": "AAD//wIVM5m/vjBvRHii2F3VR++9TIzF3bXHAA==",
      "CDVType": "ArrayBuffer"
    },
    "rawData": {
      "bytes": [
        2,
        1,
        2,
        27,
        255,
        255,
        255,
        2,
        21,
        51,
        153,
        191,
        190,
        48,
        111,
        68,
        120,
        162,
        216,
        93,
        213,
        71,
        239,
        189,
        76,
        140,
        197,
        221,
        181,
        199,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
        0
      ],
      "data": "AgECG////wIVM5m/vjBvRHii2F3VR++9TIzF3bXHAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=",
      "CDVType": "ArrayBuffer"
    }
  },
  "rssi": -66,
  "id": "4A:BE:5F:A4:40:A0",
  "name": null
}
```

**Comment by michaek** [comment]
I've got the same question. Wouldn't we want the BLE advertisement to have the uuid we construct it with? For both iOS and Android I'm seeing advertising show up in scans that has the information I've configured - except for the uuid.

**Comment by michaek** [comment]
I'll add to this that BLE scanning for the provided uuids never finds a peripheral, because they are advertised by newly generated uuids which it seems like we cannot determine from Javascript. The example code doesn't exercise this, because it scans for all uuids.

**Comment by barakataboujreich** [comment]
sorry i don't know why i didn't reply to @MtyldZ 
serviceuuids parameter in scan are your service uuids if you have set up a gatt server and initialized service uuids to fetch or write data to. unfortunately you will not be able to use this parameter to filter out the uuid emitted by this library.
This library simply imitate an iBeacon format. it is not a gatt server. 
if you want to detect the uuid emitted by this library, you will need to filter it out yourself from the advertisement data returned.
if you are you using [react-native-ble-manager](https://github.com/innoveit/react-native-ble-manager) then all the data needed to filter it out is existing, it is not visible for you because it is in bytes, transform it to hex string and you will be able to see your emitted data.

if you are not able to know how, i will try to find some time to write a small tutorial, but not a promise

**Comment by michaek** [comment]
Thanks, that's really helpful. I just finished getting the UUID out of the manufacturer data, based on [these lines](https://github.com/barakataboujreich/react-native-ble-advertise/blob/main/android/src/main/java/com/bleadvertise/BleAdvertiseModule.java#L140-L159), and I didn't have a chance to follow up here yet.

It seems like, conceptually, we just need to do `toHexString(toByteArray(device.manufacturerData).slice(4, 20))` to get the UUID back out. (We are using react-native-ble-plx, and the data comes to us in base64. The toByteArray we use is from react-native-quick-base64.)

**Comment by barakataboujreich** [comment]
so i assume react-native-ble-plx is returning first 4 bytes as company code and ibeacon identifier (0x02 0x15). the UUID should be right after these bytes. if you need any further help, let me know

**Comment by michaek** [comment]
Yep, that's what it looks like. Your help has been great, thanks!

**Comment by michaek** [comment]
Actually, @barakataboujreich, I do have a follow up question for this. It seems like that's how the manufacturer data is constructed when advertising on Android, but it seems like it happens internally within CLBeaconRegion on iOS, and I don't see anything that looks like our uuids. Do you know how we would identify an advertising iOS device by its uuid?

**Comment by barakataboujreich** [comment]
@michaek you are correct. on android we are able to customize the advertisement data as we want. on IOS we are forced with the iBeacon format of apple, using their internal methods, this is why i made the android look format look like IOS. we use this library in our production app to send advertisement packets to our hardware, and we had to use the same format to simplify the code.

to be honest i never used one phone for emission and another to try and detect the uuid emitted, i use it for phone and hardware(esp32, rpi, arduino, etc...)

i just tried using react-native-ble-manager and surprisingly the emission from the ios is returning data that i am not emitting, but the esp32 detects the data i am emitting, by cross referencing the mac address at the emission time.

i have no idea why the scanning library did not return the correct advertisement data, and i do not have access to two iphones to try an emission from one and scanning from another if it will be detected or same case.

but esp32 is detecting the emission as i am emitting from an ios

**Comment by michaek** [comment]
Thanks for your help! I'm not sure if I'm going to get to the bottom of this one at this point, but if I do I will put what I find here.

**Comment by barakataboujreich** [comment]
i just found this 
[beacons is react-native-plx](https://github.com/dotintent/react-native-ble-plx/wiki/=-FAQ:-Beacons)

which redirects us to another library [react-native-beacons-manager](https://github.com/MacKentoch/react-native-beacons-manager) now i do not have time today to even test this library but i tried the app he is referring to [beacon-emitter](https://apps.apple.com/us/app/beacon-emitter/id1665201240) which runs on iMac and i tried it, react-native-ble-manager on ios did not detect even the manufacturer id it's like the advertisement packets are hidden. on android it detected the manufacturer id of apple but the rest of the data is random and not making sence. i think now it is at API level of the OS that the advertisement data are being hidden

**Comment by michaek** [comment]
It was my suspicion, too, that it was hidden in the iOS libraries as part of the broader Apple push for privacy. There's probably some predictable way to id them (a hash, maybe) but my hunch is Apple might keep that closely held.

**Comment by barakataboujreich** [comment]
> It was my suspicion, too, that it was hidden in the iOS libraries as part of the broader Apple push for privacy. There's probably some predictable way to id them (a hash, maybe) but my hunch is Apple might keep that closely held.

There should be a way because the data is there, or the esp32 would not have been able to fetch it. Even rpi with nodejs is displaying the data. But the question is how to get the data correctly on either android or ios

**Comment by michaek** [comment]
You're absolutely right. bluetoothctl scan happily shows the correct manufacturer data.

**Comment by barakataboujreich** [comment]
i tried today to cheat the system and try to build the manufacturerData from scratch like i do on android, and not use the Core Location internal functions to build the region and iBeacon. but unfortunately Apple does not even allow that. it's like you have an Apple iPhone, not whatever your app is wanting it to be. So all the manufacturer stuff is Apple's data, and all the device stuff is the iPhone. The only way to emit a custom uuid is by being an iBeacon. 

I did some research and they provided a way for users to detect an iBeacon with it's uuid [Configure a Device to Range for Beacons](https://developer.apple.com/documentation/corelocation/ranging-for-beacons#Configure-a-Device-to-Range-for-Beacons) and this is your only chance to detect it. 

i do not know if there is a library that is implementing this or not, you might have to create your own library

**Comment by michaek** [comment]
Thanks for looking into it. I also ran into a dead end there.
